equational_theories/Closure.lean

```diff
@@ -138,35 +138,51 @@ def Bitset.set (b : Bitset) (n : Nat) : Bitset :=
 def Bitset.get (b : Bitset) (n : Nat) : Bool :=
   (b.toArray[n >>> 6]! >>> ((UInt64.ofNat n) &&& 63)) &&& 1 != 0
 
-def number_equations (inp : Array EntryVariant) : Std.HashMap String Nat × Array String := Id.run do
-  -- number the equations (arbitrarily) for easier processing
-  let mut eqs : Std.HashMap String Nat := {}
-  let mut eqs_order : Array String := #[]
+structure DenseNumbering (α : Type) [BEq α] [Hashable α] where
+  in_order : Array α
+  index : Std.HashMap α Nat
+
+def DenseNumbering.size {α : Type} [BEq α] [Hashable α] (num : DenseNumbering α) : Nat :=
+  num.in_order.size
+
+instance {α : Type} [BEq α] [Hashable α] : GetElem? (DenseNumbering α) α Nat (fun num a => a ∈ num.index) where
+  getElem num a h := num.index[a]
+  getElem? num a := num.index[a]?
+
+def DenseNumbering.fromArray {α : Type} [BEq α] [Hashable α] (elts : Array α) : DenseNumbering α :=
+  let index := Std.HashMap.ofList (elts.mapIdx (fun i x => (x, i.val))).toList
+  ⟨elts, index⟩
+
+def ltEquationNames (a b : String) : Bool :=
+  assert! a.startsWith "Equation"
+  assert! b.startsWith "Equation"
+  let aNum := (a.toSubstring.drop 8).toNat?.get!
+  let bNum := (b.toSubstring.drop 8).toNat?.get!
+  aNum < bNum
+
+def equationSet (inp : Array EntryVariant) : Std.HashSet String := Id.run do
+  let mut eqs : Std.HashSet String := {}
   for imp in inp do
     match imp with
     | .implication ⟨lhs, rhs⟩ =>
-      match eqs.containsThenInsertIfNew lhs eqs.size with
-      | (n, neqs) =>
-        eqs := neqs
-        unless n do eqs_order := eqs_order.push lhs
-      match eqs.containsThenInsertIfNew rhs eqs.size with
-      | (n, neqs) =>
-        eqs := neqs
-        unless n do eqs_order := eqs_order.push rhs
+      eqs := eqs.insert lhs
+      eqs := eqs.insert rhs
     | .facts ⟨satisfied, refuted⟩ =>
-      for lhs in satisfied ++ refuted do
-        match eqs.containsThenInsertIfNew lhs eqs.size with
-      | (n, neqs) =>
-        eqs := neqs
-        unless n do eqs_order := eqs_order.push lhs
-    | _ => pure ()
-  pure (eqs, eqs_order)
+      for eq in satisfied ++ refuted do
+        eqs := eqs.insert eq
+    | .unconditional eq =>
```

Comment:
Nice. After this change, `Equation1` starts showing up in the extracted outcomes for `Subgraph`, as it should.